README
Static Map
==========

This application provides a wrapper around Google static and dynamic map APIs
in a single interface. Given a list of coordinates, it builds the URL for
loading a map image from Google Static Maps. When clicked on the map image,
it loads the dynamic version of the same map, with the markers intact.


Dependencies
------------

StaticMap requires the jQuery library (any version after 1.2 will do) and
the jQuery Class plug-in (http://github.com/taylanpince/jquery-class/ ). A 
Google Maps API key is also needed for it to run properly. All Google
JavaScript libraries are loaded dynamically so they don't need to be included
separately.


Installation
------------

From your media folder, create a symlink to the dist directory under the
static_map path:

    ln -s ../lib/static_map/dist media/static_map

Make sure CSS and JS files are loaded on the templates where you will use 
Static Map:

    <style type="text/css">
        @import "{{ MEDIA_URL }}static_map/css/static_map.css";
    </style>
    
    <script type="text/javascript" src="{{ MEDIA_URL }}static_map/js/
        static_map{% if not DEBUG %}.min{% endif %}.js"></script>


Usage
-----

StaticMap class takes a selector and a dictionary of options:

    - api_key: Google API key, required for both static and dynamic maps
    
    - size: Dimensions of the map as an array
    
    - center: Central coordinates of the map, as an array
    
    - zoom: Zoom level of the map
    
    - markers: An array of markers, each identified as a dictionary:
    
        - coordinates: Coordinates of the marker, as an array
        
        - title: A title for the marker, will only show up on a dynamic map
        
        - description: A description for the marker

    - media_url: A string that will be apended to media file paths

    - maptype: A string describing the type of map. "roadmap" and "terrain"
               currently work, nothing else has been tested.

After creating the container for the map, instantiate the class with options:

    <div id="MapArea"></div>
    
    <script type="text/javascript">
    //<![CDATA[
        new core.StaticMap("#MapArea", {
            "media_url" : "{{ MEDIA_URL }}",
            "api_key" : "{{ GOOGLE_MAPS_API_KEY }}", 
            "size" : [298, 298], 
            "center" : [41.9288, -87.6315], 
            "zoom" : 10,
            "maptype" : "roadmap",
            "markers" : [
                {
                    "coordinates" : [41.9228, -87.6632],
                    "title" : "Title 1",
                    "description" : "Description 2"
                },
                {
                    "coordinates" : [42.9228, -88.6632],
                    "title" : "Title 2",
                    "description" : "Description 2"
                }
            ]
        });
    //]]>
    </script>


RELEASES
RELEASE 1.1.0ALPHA
==================

Tagged release-1.1.0ALPHA on April 14, 2009 by Taylan Pince


New Features
------------

* Added support for terrain map type (Matt Wilson)
* Made map zoom level same as the static image (Matt Wilson)

Bug Fixes
---------

* Fixed a problem with iterating over markers (Matt Wilson)

* * *

RELEASE 1.0.0BETA
=================

Tagged release-1.0.0BETA on March 24, 2009 by Taylan Pince


New Features
------------

* Class is self-contained now, only depends on jQuery and jQuery Class plug-in

* * *

RELEASE 1.0.0ALPHA
==================

Tagged release-1.0.0ALPHA on March 12, 2009 by Taylan Pince


New Features
------------

* Added StaticMap class that provides a wrapper around Google's Static and
  Dynamic Map applications in a single interface

* * *


src/static_map.js
/*
*	Static Map
*	A JS class for rendering static Google Maps and loading dynamic maps on the fly
*	
*	Requires jQuery library (http://www.jquery.com),
*   jQuery.Class plug-in (http://github.com/taylanpince/jquery-class)
*	
*	Taylan Pince (taylanpince at gmail dot com) - October 24, 2008
*/

$.namespace("core.StaticMap");

core.StaticMap = $.Class.extend({
    
    url_template : "http://maps.google.com/staticmap?key=%(api_key)&amp;center=%(center)&amp;markers=%(markers)&amp;size=%(size)&amp;format=png8&amp;maptype=%(maptype)",
    api_template : "http://www.google.com/jsapi?key=%(api_key)",
    marker_template : "%(media_path)static_map/images/markers/%(number).png",
    image_template : '<a href="javascript:void(0);" title="%(launch_copy)"><img src="%(map_url)" alt="Static Map" /><span class="map-link">%(launch_copy)</span></a>',
    loader_template : '<div class="map-loader"><img src="%(media_path)static_map/images/loader.gif" alt="Map Loader" /></div>',
    info_template : '<h3>%(title)</h3>%(description)',
    
    launch_copy : "Launch Interactive Map",
    
    selector : "",
    size : [300, 300],
    center : [0, 0],
    zoom : 14,
    maptype : "roadmap",
    markers : [],
    media_path : "",
    
    render_template : function(template, values) {
        for (val in values) {
            var re = new RegExp("%\\(" + val + "\\)", "g");
            
            template = template.replace(re, values[val]);
        }
        
        return template;
    },
    
    clean_variable : function(val, delimiter) {
        if (val.charAt(val.length - 1) == delimiter) {
            val = val.substring(0, (val.length - 1));
        }
        
        return val;
    },
    
    build_url : function() {
        var center = this.center[0] + "," + this.center[1];
        
        var markers = "";
        
        for (var m=0; m < this.markers.length; m++) {
            markers += this.markers[m].coordinates;

            if (this.markers.length > 1 || center != this.markers[m].coordinates) {
                markers += ",red" + (parseInt(m) + 1) + "|";
            }
        }
        
        var url = this.render_template(this.url_template, {
            "api_key" : this.api_key,
            "size" : this.size[0] + "x" + this.size[1],
            "center" : center,
            "markers" : this.clean_variable(markers, "|"),
            "maptype" : this.maptype
        });
        
        url += "&amp;zoom=" + this.zoom;
        
        return url;
    },
    
    draw_static_map : function() {
        $(this.selector).html(this.render_template(this.image_template, {
            "map_url" : this.build_url(),
            "launch_copy" : this.launch_copy
        })).find("a").click(this.launch_interactive_map.bind(this));
    },
    
    draw_interactive_map : function() {
        var gmap = new google.maps.Map2($(this.selector).get(0));
        var gcenter = new google.maps.LatLng(this.center[0], this.center[1]);
        var gbounds = new google.maps.LatLngBounds();
        
        gbounds.extend(gcenter);
        gmap.setCenter(gcenter);
        gmap.addControl(new google.maps.SmallMapControl());
        
        for (var m = 0; m < this.markers.length; m++) {
            var gpoint = new google.maps.LatLng(this.markers[m].coordinates[0], this.markers[m].coordinates[1]);
            var gmarker = new google.maps.Marker(gpoint, {
                "icon" : new google.maps.Icon(google.maps.DEFAULT_ICON, this.render_template(this.marker_template, {
                    "media_path": this.media_path,
                    "number": ((this.markers.length == 1 && gcenter.equals(gpoint)) ? "dot" : (parseInt(m) + 1))
                }))
            });
            
            if (typeof this.markers[m].title || typeof this.markers[m].description) {
                gmarker.bindInfoWindowHtml(this.render_template(this.info_template, {
                    "title" : this.markers[m].title,
                    "description" : this.markers[m].description
                }), {
                    "maxWidth" : (parseInt(this.size[0]) - 100)
                });
            }
            
            gmap.addOverlay(gmarker);
            gbounds.extend(gpoint);
        }
        
        gmap.setZoom(this.zoom);

        if (this.maptype == "terrain") {
            gmap.addMapType(G_PHYSICAL_MAP);
            gmap.setMapType(G_PHYSICAL_MAP);
        }

    },
    
    load_interactive_map : function() {
        if (typeof google == "object") {
            google.load("maps", "2", {
                "callback" : this.draw_interactive_map.bind(this)
            });
        } else {
            this.draw_static_map();
            
            alert("There was an error while loading the interactive map. We are sorry for the inconvenience.");
        }
    },
    
    launch_interactive_map : function() {
        this.show_loader();
        
        if (typeof google == "object") {
            this.load_interactive_map();
        } else {
            $.getScript(this.render_template(this.api_template, {
                "api_key": this.api_key
            }), this.load_interactive_map.bind(this));
        }
        
        return false;
    },
    
    show_loader : function() {
        $(this.selector).append(this.render_template(this.loader_template, {
            "media_path": this.media_path
        }));
    },
    
    hide_loader : function() {
        $(this.selector).find(".map-loader").remove();
    },
    
    init : function(selector, options) {
        this.selector = selector;
        
        $(this.selector).addClass("static-map");
        
        if (options) {
            for (opt in options) {
                this[opt] = options[opt];
            }
        }
        
        this.draw_static_map();
    }
    
});
